Data: Add a util to read write partition stats

ajantha-bhat opened the pull request:
Based on the conclusion from https://github.com/apache/iceberg/pull/9437, we are going with the local implementation with multi threads instead of distributive algorithm. 

`iceberg-core` module cannot write stats as it doesn't depend on parquet code. 
So, `iceberg-data` module is introducing an API to write it and core module just registers it to the table. 

Syntax look like this.
```
PartitionStatisticsFile result = PartitionStatsHandler.computeAndWritePartitionStatsFile(table);
table.updatePartitionStatistics().setPartitionStatistics(result).commit();
```

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java:
```diff
@@ -0,0 +1,107 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GeneratePartitionStats {
```
`iceberg-core` module cannot write stats as it doesn't depend on parquet code.
So, `iceberg-data` module is introducing an API to write it and core module just registers it to the table.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+        Types.NestedField.required(3, Column.DATA_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.required(4, Column.DATA_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.required(5, Column.DATA_FILE_SIZE_IN_BYTES.name(), Types.LongType.get()),
+        Types.NestedField.optional(
```
observe optional and required fields, as per spec.

ajantha-bhat replied:
last computed snapshot may be unavailable at the time of computing the stats (due to expire snapshots). So, `LAST_UPDATED_SNAPSHOT_ID` and `LAST_UPDATED_AT` can be null. Hence, it is optional. Also, it doesn't make sense to fill it with 0 as snapshot id 0 is not meaningful in iceberg.

To maintain uniform behaviour of all optional fields, the stats like `TOTAL_RECORD_COUNT, EQUALITY_DELETE_FILE_COUNT, EQUALITY_DELETE_RECORD_COUNT, POSITION_DELETE_FILE_COUNT, POSITION_DELETE_RECORD_COUNT` will be made null when absent instead of 0.

I have added a nice testcase which validates all these now. 

Also, schema(optional & required) is as spec: https://iceberg.apache.org/spec/#partition-statistics-file
If we decide to change optional to required now, it has to go through vote :(

aokolnychyi replied:
I understand why we mark `LAST_UPDATED_SNAPSHOT_ID` and `LAST_UPDATED_AT ` as optional. My question was about counts. In my view, null means unknown and 0 means absent. If I read a partition stats file and see null, it won't be informative.

aokolnychyi replied:
It makes sense to keep fields optional but I would expect implementations to write 0 for the position delete record count if a partition has no deletes.

ajantha-bhat replied:
done.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    record.set(Column.SPEC_ID.ordinal(), entry.file().specId());
+
+    Snapshot snapshot = table.snapshot(entry.snapshotId());
+    if (snapshot != null) {
```
if the snapshot is expired, last updated time and last updated snapshot will be null.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+            "Unsupported file content type: " + entry.file().content());
+    }
+
+    // TODO: optionally compute TOTAL_RECORD_COUNT based on the flag
```
the idea of this field is to apply the equality or position delete filters and give the effective results. Need to add a logic to compute it in the follow up.

RussellSpitzer replied:
That would be tricky, not sure we can actually do it without actually scanning rows.

ajantha-bhat replied:
True. But we decided we need to have this optional field during spec design. 

Plan is to have a follow up PR with a flag that enables computing this. By default it won't be computed as it is an expensive operation.

aokolnychyi replied:
I am not sure we would want to extend this path to compute deletes. We may also reconsider how position deletes are handled in V3, so it may be possible to compute this without scanning the data.

I'd probably drop this TODO for now.

ajantha-bhat replied:
True. Maybe we didn't think well about this field during spec design. We can drop a todo and make this field deprecated if required later on.

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,106 @@
+        GenericFileWriterFactory.builderFor(table)
+            .dataSchema(dataSchema)
+            .dataFileFormat(fileFormat)
+            .dataSortOrder(sortOrder(dataSchema))
```
sorting as per the unified partition tuple order as per the spec.

lirui-apache replied:
I don't think this really does the sorting. Seems it only marks the generated file with the sort ID?

RussellSpitzer replied:
Yeah I'm not sure this actually has any use, not sure

ajantha-bhat replied:
I see. I assumed it will sort the data as we are passing to writer. Looks like it just adds sort order to metadata. 

During Spec design phase, we decided that it is good to sort the partition stats before writing to a file so that it can help for faster lookup and better compression. But seems like we need engine to sort it. Just the Java API of parquet writers we have doesn't have the option to sort it?

So, we should update back the spec that these fields cannot be sorted in a separate PR. WDYT?

lirui-apache replied:
How about using a sorted map to store the stats entries? I suppose we can get a comparator for the partition data using the `Comparators.forType` util, right?

ajantha-bhat replied:
done

ajantha-bhat commented:
ping @aokolnychyi

ajantha-bhat commented:
cc: @RussellSpitzer, @szehon-ho : Can you please take a look? I need reviews and I want to take it forward.
previous discussions can be found at https://github.com/apache/iceberg/pull/9437

lirui-apache commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  public static void updateRecord(Record toUpdate, Record fromRecord) {
+    toUpdate.set(
```
Does this mean we allow merging records with different spec IDs? Suppose we start with `Spec1: {Field#1}`. And we write a partition `Filed#1 = 1`. Then we change spec to `Spec2: {Field#1, Field#2}`, and write another partition `Field#1 = 1 / Field#2 = null`. Will we only have one entry in the partition stats file with `Partition_Data: {Field#1 = 1 / Field#2 = null}, Spec_ID: 2`?

ajantha-bhat replied:
yes. please refer to the last point of https://github.com/apache/iceberg/pull/9437#issuecomment-2064341275

Also, it is similar to how the partitions metadata table implemented currently.

lirui-apache replied:
Hmm.. but how can such a stats entry be useful? From a user's perspective, there's no way to tell if an entry is merged. Suppose the entry says we have 10 records for partition `{Field#1 = 1 / Field#2 = null}`. This doesn't mean there're 10 records satisfying filter `Field#1 = 1 and Field#2 is null`, right?

ajantha-bhat replied:
lets us go with the example, 
partitioned with `col1` and inserted 5 rows for that partition `col1 = x`. 
now the spec has evolved with partition as `col1/col2` and inserted 3 rows for `col1 = x` and `col2 = y`
and inserted 7 rows for `col1 = x` and `col2 = null`

So, when the query is run with `col1 = x` and `col2 = null`, it will fetch 12 (5+7) rows. So, the partition stats will also have stats from 12 rows for `col1 = x` and `col2 = null` (unified tuple). So, I think it is fine. 

Can you elaborate why do you think it is a problem to have unified tuple?

lirui-apache replied:
We're planning to use partition level stats to accelerate queries fetching partition metadata. I think the unified tuple can be confusing for users. For example, if the query says there're 12 rows in partition `col1 = x/col2 = null`, a user cannot tell whether we inserted 12 rows into that partition with spec2, or we inserted 5 and 7 rows with spec1 and spec2 respectively.

So maybe my question should be whether the partition stats are only intended for query optimizers, and not suitable for our use case? I thought it's user-facing because it has information such as `last_updated_at`, which optimizers usually don't care about.

lirui-apache replied:
@ajantha-bhat Thanks for clarifying! Personally I still think this feature can be used to improve user experience for investigating partitions. Maybe we just need to explain to users about the ambiguity for partitions with null values.

ajantha-bhat replied:
@lirui-apache: You were right that we shouldn't combine the stats for different spec. Just that we didn't had a strong usecase. 

Anton mentioned that 
> If we combine delete stats across specs, they will not be accurate. A delete file from one spec never applies to a data file in another spec.

So, now we are not combining it. 

more details on https://github.com/apache/iceberg/pull/11146

lirui-apache replied:
OK, thanks for the pointer

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+  private PartitionStatsUtil() {}
+
+  public enum Column {
+    PARTITION_DATA,
```
nit: Just called "partition" in the Spec

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    SPEC_ID,
+    DATA_RECORD_COUNT,
+    DATA_FILE_COUNT,
+    DATA_FILE_SIZE_IN_BYTES,
```
called: total_data_file_size_in_bytes in the spec

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    CloseableIterable<? extends ManifestEntry<? extends ContentFile<?>>> entries =
+        CloseableIterable.transform(
+            ManifestFiles.open(manifest, table.io(), table.specs())
+                .select(scanColumns(manifest.content())) // don't select stats columns
```
I would remove these comments

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+                    t.copyWithoutStats());
+
+    return CloseableIterable.transform(
+        entries, entry -> fromManifestEntry(entry, table, recordSchema));
```
Why did we need to do a copy if we are going to create a new object and discard the manifest entry here?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+        entries, entry -> fromManifestEntry(entry, table, recordSchema));
+  }
+
+  public static void updateRecord(Record toUpdate, Record fromRecord) {
```
Method name here is pretty vague, also anything public should have a javadoc

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+  private static void incrementLong(Record toUpdate, Record fromRecord, Column column) {
+    toUpdate.set(
+        column.ordinal(),
+        (long) toUpdate.get(column.ordinal()) + (long) fromRecord.get(column.ordinal()));
```
Since record is structlike we can do get(column.ordinal(), Long.class)
might look a little cleaner

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  private static void checkAndIncrementLong(Record toUpdate, Record fromRecord, Column column) {
+    if (toUpdate.get(column.ordinal()) != null && fromRecord.get(column.ordinal()) != null) {
```
If toUpdate is null but fromRecord is not null shouldn't we set toUpdate to fromRecord?

RussellSpitzer replied:
I also think we don't gain a lot by splitting out the check here, do we have a usecase for incrementLong without checking?

ajantha-bhat replied:
> I also think we don't gain a lot by splitting out the check here, do we have a usecase for incrementLong without checking?

There are some optional fields and some required fields, I wanted to check null only for optional fields. Hence I had two separate methods to avoid null check for required fields.  But it may not matter much. I will simplify.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    checkAndIncrementLong(toUpdate, fromRecord, Column.EQUALITY_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toUpdate, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (toUpdate.get(Column.LAST_UPDATED_AT.ordinal()) != null
```
Same question as below, shouldn't we set toUpdate to the value of fromRecord if toUpdate is null but fromRecord is not?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    return data;
+  }
+
+  private static List<String> scanColumns(ManifestContent content) {
```
Should this be in BaseScan? This may be a fine location as well just wondering if it belongs here

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java:
```diff
@@ -0,0 +1,107 @@
+        .executeWith(ThreadPools.getWorkerPool())
+        .onFailure(
+            (file, thrown) ->
+                LOG.warn(
```
We aren't going to fail if the manifest is broken?

ajantha-bhat replied:
stopping on failure and throwing exception now. Still keeping warning to print the file path (like other places of code)

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestComputePartitionStats.java:
```diff
@@ -0,0 +1,155 @@
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class TestComputePartitionStats {
```
Needs tests covering the null behaviors I mentioned above

RussellSpitzer replied:
I think we should have some tests

1. Tests where the snapshot is empty
2. Tests where some of stats are null
3. Tests where the partition spec has changed for some of the files

ajantha-bhat replied:
Added all these. Will add some more by tomorrow for equality and position delete stats and validation for snapshot ids etc.

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestPartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,122 @@
+
+      partitionListBuilder.add(record);
+    }
+    List<Record> records = partitionListBuilder.build();
```
nit: expectedRecords

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestPartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,122 @@
+
+    Assertions.assertThat(Paths.get(outputFile.location())).exists();
+
+    List<Record> rows;
```
nit: writtenRecords

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestPartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,122 @@
+      rows = Lists.newArrayList(recordIterator);
+    }
+
+    Assertions.assertThat(rows).hasSize(records.size());
```
This whole block can use the assertIterableEquals
```java
 Assertions.assertIterableEquals();
```

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestPartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,122 @@
+    ImmutableList.Builder<Record> partitionListBuilder = ImmutableList.builder();
+
+    long seed = System.currentTimeMillis();
+    LOG.info("Seed used for random generator is {}", seed);
```
not sure this will be useful

RussellSpitzer commented on data/src/test/java/org/apache/iceberg/data/TestPartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,122 @@
+
+    ImmutableList.Builder<Record> partitionListBuilder = ImmutableList.builder();
+
+    long seed = System.currentTimeMillis();
```
In general, unless we are doing an explicit property test we tend to make sure the tests are consistent. So I would just freeze the seed as a constant in the file.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,213 @@
+    }
+  }
+
+  public static Record partitionDataToRecord(
```
Because the parquet writer expects an Iceberg `Record` and `IndexedRecord` is not an Iceberg `Record`

```
java.lang.ClassCastException: class org.apache.iceberg.PartitionData cannot be cast to class org.apache.iceberg.data.Record (org.apache.iceberg.PartitionData and org.apache.iceberg.data.Record are in unnamed module of loader 'app')
	at org.apache.iceberg.data.parquet.GenericParquetWriter$RecordWriter.get(GenericParquetWriter.java:41)
	at org.apache.iceberg.parquet.ParquetValueWriters$StructWriter.write(ParquetValueWriters.java:580)
	at org.apache.iceberg.parquet.ParquetValueWriters$StructWriter.write(ParquetValueWriters.java:581)
	at org.apache.iceberg.parquet.ParquetWriter.add(ParquetWriter.java:135)
	at org.apache.iceberg.io.DataWriter.write(DataWriter.java:71)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.iceberg.data.PartitionStatsWriterUtil.writePartitionStatsFile(PartitionStatsWriterUtil.java:82)
```

lirui-apache replied:
Maybe there're other data types we need to convert. E.g. the generic writer for `Date` expects a `LocalDate`, while the internal class for `Date` is an integer.

ajantha-bhat replied:
I will take a look. Is there a util class to handle this already? Because isn't it a common problem?

lirui-apache replied:
I believe `IdentityPartitionConverters` can be used to do the conversion. And perhaps we can have test cases for date/time partition columns?

ajantha-bhat replied:
@lirui-apache: You are right about needing extra conversion for few data types. I have added the test case for all the data type as partition now in `TestPartitionStatsWriterUtil#testAllDatatypePartition` and also improved `IdentityPartitionConverters`

ajantha-bhat replied:
Good suggestion. Optimized it.

ajantha-bhat commented:
@RussellSpitzer: Thanks for the review. 
I have addressed all the comments and added the mentioned tests. I have plan to add some more tests tomorrow (related to delete file stats)

aokolnychyi commented:
I should finally have some time today. @ajantha-bhat, is PR #9437 still valid as well?

aokolnychyi commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+        record.set(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.ordinal(), entry.file().fileSizeInBytes());
+        break;
+      case POSITION_DELETES:
+        record.set(Column.POSITION_DELETE_FILE_COUNT.ordinal(), 1);
```
If there are no deletes, what kind of values will we have as delete counts? Null or 0?

ajantha-bhat replied:
Null, as mentioned in https://github.com/apache/iceberg/pull/10176#discussion_r1698066417, 

also added the tests.

aokolnychyi reviewed (commented):
I think this is going into the right direction. I am sorry it took so long to get back, I have more time now to look into this.

I did a detailed pass but there are a few points that I'll have to think about. It is also super critical to add a JMH benchmark to guide us what's best in a few places.

lirui-apache commented on data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java:
```diff
@@ -0,0 +1,122 @@
+    // Map of partitionData, partition-stats-entry per partitionData.
+    // Sorting the records based on partition as per spec.
+    Map<Record, Record> partitionEntryMap =
+        new ConcurrentSkipListMap<>(Comparators.forType(partitionType));
```
I also tried ConcurrentSkipListMap, but it seems its `compute` and `merge` methods are not guaranteed to be atomic, which can lead to consistency issues. So I guess we need to sort the records after we populate the map.

ajantha-bhat replied:
@lirui-apache, You are right. I pushed a new test now which validates each and every stats. I observed flakiness because of  `ConcurrentSkipListMap#compute`. 

I will sort after populating the map as suggested by you and Anton. JMH, I will add in a follow up PR. I think we can incrementally improve this logic based on JMH benchmarks later on.

ajantha-bhat replied:
Added the JMH benchmarks, for 30K manifests (1 add, 2 update operation per partition value * 10k partitions), it took only 0.5 seconds with concurrent hashmap. This includes reading 30k manifests in multi threaded way and computing the concurrent, sorting the map and writing the parquet partition stats and committing to the table. 

I wanted to try 100K partitions but the problem is generating 30k manifests and committing them to table took around 30 minutes. So, it takes lot of setup time if I increase the manifests. 

Since 0.5 seconds is very small. Do we still need to validate concurrent hashmap vs Threadlocal hashmap performance? 


```
Benchmark                                             Mode  Cnt  Score   Error  Units
PartitionStatsGeneratorBenchmark.writePartitionStats    ss    5  0.556 ± 0.343   s/op
```

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+   */
+  public static Schema schema(Types.StructType partitionType) {
+    if (partitionType.fields().isEmpty()) {
+      throw new IllegalArgumentException("getting schema for an unpartitioned table");
```
This is unified partitionType. If the table is currently unpartitioned but it was partitioned before, we still allow computing stats. 

Yes. If the table is never partitioned, we are not allowing computing stats because it is of no use (or it is not a partition stats anymore).

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+    return CloseableIterable.transform(
+        ManifestFiles.open(manifest, table.io(), table.specs())
+            .select(BaseScan.scanColumns(manifest.content()))
+            .liveEntries(),
```
I wanted to avoid duplicate stats, for example when compacted or for rewrite manifests, old manifests will be marked as `DELETED`. Hence, keeping liveEntries will help skipping those manifests. 

Also, if the manifests are marked as `DELETED`, it won't be considered for table scan. So, since these stats are used for query planning, I think it is safe to skip these. Thoughts?

aokolnychyi replied:
The problem is that the computed `last_updated_at` may no longer be accurate (we won't catch operations that modified partitions by removing data files). We can project the entry status and only take deleted entires into account when computing the last updated timestamp. Does it worth the extra complexity? We'll have to think.

> compacted or for rewrite manifests, old manifests will be marked as DELETED

Don't we actually remove such manifests from the manifest list?

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+   */
+  public static Schema schema(Types.StructType partitionType) {
+    if (partitionType.fields().isEmpty()) {
+      throw new IllegalArgumentException("getting schema for an unpartitioned table");
```
updated.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+        Math.max(
+            (int) toRecord.get(Column.SPEC_ID.ordinal()),
+            (int) fromRecord.get(Column.SPEC_ID.ordinal())));
+    checkAndIncrementLong(toRecord, fromRecord, Column.DATA_RECORD_COUNT);
```
Can change it to use `Enum.name()` instead of ordinal with `getField()` and `setField()`.  But `getField` returns `Object` and needs explicit casting. Hence, Russell also suggested to use get() with classType to avoid casting in some comment long back.

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java:
```diff
@@ -0,0 +1,122 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GeneratePartitionStats {
```
Since I started as a spark action, used verb. Will change it to noun.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,227 @@
+    checkAndIncrementInt(toRecord, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (fromRecord.get(Column.LAST_UPDATED_AT.ordinal()) != null) {
+      if (toRecord.get(Column.LAST_UPDATED_AT.ordinal()) == null
```
> Hm, can we actually compute last_updated_at if at least one side has null? 

`fromRecord` side if it is null, then that snapshot might have expired and older than `toRecord` side. So, it won't be latest. Hence, no need to update. 

If `toRecord` side is null then `fromRecord` might be latest hence updating. 

So, when we go through all records for that partition, it will have the latest snapshot info.

aokolnychyi replied:
If we assume expired snapshots are older (which is mostly true unless someone expires a snapshot in the middle of the chain), then this logic seems to work.

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/GeneratePartitionStats.java:
```diff
@@ -0,0 +1,122 @@
+    Snapshot currentSnapshot = SnapshotUtil.latestSnapshot(table, branch);
+    if (currentSnapshot == null) {
+      Preconditions.checkArgument(
+          branch == null, "Couldn't find the snapshot for the branch %s", branch);
```
No. Empty branches will have a dummy snapshot without manifest files. 

https://github.com/apache/iceberg/blob/76dba8fe83c4496318cc34436e610cf50f43054d/core/src/main/java/org/apache/iceberg/SnapshotManager.java#L81

So, I am not throwing exception for that case. 
Added test and It creates an empty stats file for the dummy snapshot since snapshot exist.

ajantha-bhat commented:
Thanks @RussellSpitzer, @aokolnychyi and @lirui-apache for the review feedbacks. 
I have also added strong test validations today.

I have addressed or replied to each comment. 

As per me, pending things are
a) https://github.com/apache/iceberg/pull/10176#discussion_r1691044481, need to check about record converter and more unit tests for record converter for each data type. 
b) JMH benchmark for concurrent map logic to try out performance optimizations. 
I will check these tomorrow. 

Let me know if you can do another pass. Lets not lose momentum on this again.

lirui-apache commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,93 @@
+  public static void writePartitionStatsFile(
+      Table table, Iterator<Record> records, OutputFile outputFile) {
+    Schema dataSchema = PartitionStatsUtil.schema(Partitioning.partitionType(table));
+    FileFormat fileFormat =
```
The reader only supports Parquet at the moment. Why we make the writer support different formats?

ajantha-bhat replied:
For writer generic writer was available. For reader it is not available. Hence, it is a TODO. 
I have mentioned the same at https://github.com/apache/iceberg/pull/10176#discussion_r1571025684

The idea is to have a generic reader in a separate PR and enhance this feature to use it.

lirui-apache reviewed (commented):
@ajantha-bhat Thanks for updating. I only have some minor comments, otherwise looks good to me.

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsGenerator.java:
```diff
@@ -0,0 +1,128 @@
+
+  public PartitionStatsGenerator(Table table, String branch) {
+    this.table = table;
+    this.branch = branch;
```
branch or tags are finally mapped to snapshot. So, Once the query engine resolves the branch/tag to a particular snapshot, it can check/compute the stats for that snapshot and track it from table metadata.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,240 @@
+
+    return new Schema(
+        Types.NestedField.required(1, Column.PARTITION.name(), partitionType),
+        Types.NestedField.required(2, Column.SPEC_ID.name(), Types.IntegerType.get()),
```
The intention is to indicate the latest spec id that has modified this partition value. 

After the partition evolution, even though we use unified tuple, if the partition is not modified by the latest spec id, It will still be remained as old spec id value. 

This can help in knowing how many partitions have the new data inserted as per new spec and how many partition might need to call rewrite data files (to convert them to new spec)

ajantha-bhat commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsWriterUtil.java:
```diff
@@ -0,0 +1,93 @@
+      Table table, Iterator<Record> records, OutputFile outputFile) {
+    Schema dataSchema = PartitionStatsUtil.schema(Partitioning.partitionType(table));
+    FileFormat fileFormat =
+        FileFormat.fromString(
```
I didn't get it fully. 

Do you mean, we should also store the output file format type in the `PartitionStatisticsFile` for the reader to be known? Currently based on the file extension, we can know which reader to be used. Hence, it is not explicitly stored.

ajantha-bhat commented:
ping @aokolnychyi, @RussellSpitzer

lirui-apache reviewed (approved):
LGTM, thanks @ajantha-bhat for updating

ajantha-bhat commented:
Thanks a lot for the review @lirui-apache, 

I hope we can get this in after a review from @RussellSpitzer and @aokolnychyi.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  /**
+   * Generates a schema as per partition statistics spec based on the given partition type.
```
Nit here and other places in the doc, the doc string is probably clearer as 
"Generates the Partition Stats Files Schema based on a given a partition type" 

Should we also note this should be called with the unified spec? Or is it ok to use with a non-unified spec?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,218 @@
+    GenericRecord record = GenericRecord.create(recordSchema);
+    Types.StructType partitionType =
+        recordSchema.findField(Column.PARTITION.name()).type().asStructType();
+    Record partitionData = coercedPartitionData(entry.file(), table.specs(), partitionType);
```
Does this work with an unpartitioned  spec? or are we guarding against it elsewhere?

ajantha-bhat replied:
Currently `recordSchema` was computed by guarded by `PartitionStatsUtil.schema()` and passed to this method. Hence, it was guarded. 

But as it is a public API, better to add the check here also. I will update it.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,218 @@
+            "Unsupported file content type: " + entry.file().content());
+    }
+
+    // Note: Not computing the `TOTAL_RECORD_COUNT` for now as it needs scanning the data.
```
Couldn't we do this as long as there are no Position or Equality Deletes? Since we are doing an aggregation we could do something like adding up the total record count but then setting it to null if we ever see a position or equality delete? 

Could be a follow up

ajantha-bhat replied:
when the position or equality deletes are not present, we can just use `DATA_RECORD_COUNT` field as it will be same as expected `TOTAL_RECORD_COUNT` for the given partition value.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  private static void checkAndIncrementLong(Record toUpdate, Record fromRecord, Column column) {
+    if (fromRecord.get(column.ordinal()) != null) {
```
Personal prefrence from me here but I find it clearer (albeit longer) to do
```java
if (from !=null && to !=null) {
   to = from + to
} else if (from != null) {
   to = from;
}
```

Just to decrease the nesting

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/data/IdentityPartitionConverters.java:
```diff
@@ -48,6 +50,13 @@ public static Object convertConstant(Type type, Object value) {
         if (value instanceof GenericData.Fixed) {
           return ((GenericData.Fixed) value).bytes();
+        } else if (value instanceof ByteBuffer) {
+          return ((ByteBuffer) value).array();
```
Use  
```java
ByteBuffers.toByteArray
```
https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/api/src/main/java/org/apache/iceberg/util/ByteBuffers.java#L27

A safer option (same suggestion for UUID)

RussellSpitzer commented on data/src/jmh/java/org/apache/iceberg/PartitionStatsGeneratorBenchmark.java:
```diff
@@ -0,0 +1,131 @@
+  private String baseDir;
+
+  // create 3 manifests with same partition values
+  private static final int ITERATION_COUNTER = 3;
```
Why is it called iteration counter? Just a little confusing since we are in a benchmark class which already defines iterations as the number of times we run the test?

RussellSpitzer commented on data/src/jmh/java/org/apache/iceberg/PartitionStatsGeneratorBenchmark.java:
```diff
@@ -0,0 +1,131 @@
+  private static final int PARTITION_COUNT = 10000;
+
+  // one data file manifest
+  private static final int DATA_FILES_PER_PARTITION_COUNT = 1;
```
Seems like this would be very unrealistic for benchmarking? Shouldn't we have multiple data files in each manifest (also multiple partitions in the same manifest?)

ajantha-bhat replied:
It takes lot of time to generate manifests (for 10k manifests it took around 20 minutes). 
I will optimize it for a balanced number and user can increase the counters if they want to see benchmarks against more manifests with less data files or less manifests with more data file per manifests.

ajantha-bhat replied:
Updated now.

2 million data file entires. 

100 partitions * 20 data files per partitions = 2k data files per manifest. 
10K manifests. 

It took around 10 seconds to write, commit and read the partition stats file.

RussellSpitzer commented on data/src/jmh/java/org/apache/iceberg/PartitionStatsGeneratorBenchmark.java:
```diff
@@ -0,0 +1,131 @@
+  @TearDown
+  public void tearDownBenchmark() throws IOException {
+    // validate row count
+    try (CloseableIterable<Record> recordIterator =
```
I don't think this is an appropriate place to assert that the benchmark ran correctly, one issue here is a failed run will not clean up the temp folder. 

Is there a reason to not place this in the benchmark itself?

ajantha-bhat replied:
Wanted to benchmark just write performance without any read code. But you are right. I have to move it.

ajantha-bhat commented on data/src/jmh/java/org/apache/iceberg/PartitionStatsGeneratorBenchmark.java:
```diff
@@ -0,0 +1,131 @@
+  // create 3 manifests with same partition values
+  private static final int ITERATION_COUNTER = 3;
+
+  // create 3 * 10000 manifests
```
Included the ITERATION_COUNTER, I will reword as 10000 manifest per iteration.

RussellSpitzer replied:
Iteration probably shouldn't be used as a name here since it conflates with the benchmark itself. I would look for another  variable name.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/PartitionStatsGenerator.java:
```diff
@@ -0,0 +1,148 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PartitionStatsGenerator {
```
Could we doc this and Writer Util? I'm mostly interested in the separation of these two classes and what should their responsibilities be?

ajantha-bhat replied:
Updated the javadocs. 

`PartitionStatsGenerator` is to generate the stats file using util classes. Util class is a helper for that.

ajantha-bhat commented:
Thanks @RussellSpitzer for the review. I have updated the comment.

aokolnychyi commented on data/src/jmh/java/org/apache/iceberg/PartitionStatsGeneratorBenchmark.java:
```diff
@@ -0,0 +1,135 @@
+        Paths.get(new File(System.getProperty("java.io.tmpdir")).getAbsolutePath()).toString();
+    table = TestTables.create(new File(baseDir), "foo", SCHEMA, SPEC, SortOrder.unsorted(), 2);
+
+    IntStream.range(0, MANIFEST_COUNTER)
```
It would be nice to parallelize the preparation.

aokolnychyi commented:
I took a detailed look at the implementation/spec today. Here are my high-level points:

- The local algorithm was the right choice. I am on board with using `ConcurrentHashMap` and coerced partitions as keys.
- The logic is split between 3 utility classes that are hard to navigate.
- In my view, `Record` is not the best abstraction to expose when computing stats in `core`.
    - Projecting and getting values from `Record` is very fragile.
    - It is necessary to have `Record` to use generic writers but `PartitionStatsUtil` is in `core` and is supposed to be helpful for arbitrary implementations, not only those that rely on generic writers.
    - The `fromManifest` method in `PartitionStatsUtil` does not return a valid `Record` as it doesn’t transform values (it shouldn’t, the writer has to do that). We need a better abstraction in `core`.

I'd consider restructuring the code as below (names are suggestions):

1. Add a wrapper class to represent partition stats in `core`.

```
public class PartitionStats {

  private final int specId;
  private final StructLike partition;
  ...
  private Long lastUpdatedSnapshotId = null;

  public void liveEntry(ContentFile<?> file, Snapshot snapshot) {
    // move the existing logic for updating stats from live entires here
  }

  public void deletedEntry(Snapshot snapshot) {
    if (snapshot != null) {
      // update last modified info to have accurate stats
    }
  }
}
```

2. Keep `PartitionStatsUtil` in `core` but change its purpose. It does everything except writing.

```
public class PartitionStatsUtil {

  public static final int PARTITION_FIELD_ID = 1;
  public static final String PARTITION_FIELD_NAME = "partition";
  public static final NestedField SPEC_ID = required(2, "spec_id", IntegerType.get());
  ...
 
  public static Schema schema(StructType partitionType) {
    // return the schema of records in partition stats files
  }

  public static Iterable<PartitionStats> computeStats(Table table, Snapshot snapshot) {
    // compute stats like in this PR using ConcurrentHashMap
    // don't convert ManifestEntry to PartitionStats, invoke liveEntry/deletedEntry to update stats as needed
  }

  public static List<PartitionStats> sortStats(Iterable<PartitionStats> stats, StructType partitionType) {
    // sort stats per spec if they have to be written (having this separately is optional)
  }
}
```

3. Add `PartitionStatsWriter` that would wrap `PartitionStats` as `Record` internally and write using generic writers (like in this PR). The writer will be responsible for converting values using `IdentityPartitionConverters`.

What do you think, @ajantha-bhat?

ajantha-bhat commented:
> Add a wrapper class to represent partition stats in core.

I didn't use the wrapper class initially and directly used record because for each entry having two objects (wrapper and Record) might increase memory footprint. But looking at benchmarks, it may not be a deal breaker and it will look clean as you suggested. 

I will update the code as you suggested.

ajantha-bhat commented:
> Add PartitionStatsWriter that would wrap PartitionStats as Record

This abstraction isn't working out. 
Because `PartitionStats` need partition as `PartitionData` and `Record` need partition as `Record`. 
Converting them on wrapping and unwrapping is becoming very messy because storage is same and there are multiple paths to store it. 

I am thinking to introduce `PartitionStatsRecord` that extends `Record` and store it in `Core` similar to `GenericRecord` and make the interfaces based on that in `core` and `data` instead of `PartitionStats`. 

Thoughts?

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/data/PartitionStatsRecord.java:
```diff
@@ -0,0 +1,172 @@
+import org.apache.iceberg.types.Types.StructType;
+
+public class PartitionStatsRecord implements Record, StructLike {
+  private static final LoadingCache<StructType, Map<String, Integer>> NAME_MAP_CACHE =
```
Similar to existing `GenericRecord` class, but it wraps the existing `PartitionStats` and used only for partition stats

aokolnychyi commented:
Let me take a look hopefully over weekend or on Monday.

aokolnychyi commented:
Okay, I played around with this a bit over the weekend. I suggest we split this PR into two parts: `PartitionStatsUtil` that computes and sorts the stats (1) and the conversion and writing of the stats into a file (2). The reason for the split is that we may need to reconsider the writing a bit. The current problem is the need to translate to generic record for writing into Parquet. I just learned that there is an ongoing effort to support writes with the internal Iceberg object model, which will nicely solve this problem for us (given that it is already in progress, it shouldn't delay this work).

While we look into the writing part, it would be nice to get `PartitionStatsUtil` in. For that, I would suggest making some tweaks to `PartitionStats` and `PartitionStatsUtil`:

- Use `StructLike` instead of `Record`.
- Don't create a temporary `PartitionStats` from a manifest entry to simply merge it with others.

ajantha-bhat commented:
> I just learned that there is an ongoing effort to support writes with the internal Iceberg object model, which will nicely solve this problem for us (given that it is already in progress, it shouldn't delay this work).

Could you please point me to that work?
